Port prelu to ATen

ezyang opened the pull request:
This operator is a warmup I was doing before tackling convolution, as it
has many properties that make it a "first" for implementing things.  In
particular, it is the first operator whose backwards have multiple
returns; this means its double backwards is the first backwards for a
function with multiple differentiable outputs.  This exercises new code
for output_mask and set_flags.

There's also some documentation and small feature additions to ATen.

ezyang committed:
Add operator<< for at::Type

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
More elaborate error message when expand fails.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Add expand_as/type_as to ATen.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Make prelu an ATen op.

This operator is a warmup I was doing before tackling convolution, as it
has many properties that make it a "first" for implementing things.  In
particular, it is the first operator whose backwards have multiple
returns; this means its double backwards is the first backwards for a
function with multiple differentiable outputs.  This exercises new code
for output_mask and set_flags.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Some developer notes for ATen.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Make 'name' occur first in Declarations.yaml

Whenever I used to read Declarations.yaml, it would drive me batty that
'name' was always embedded somewhere in the middle of the record.
Now it at the top, as it should be!

What it looks like now:

  - name: storage_offset
    method_prefix: m_
    arguments:
    - dynamic_type: Tensor
      name: self
      type: const Tensor &

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Only set_flags on differentiable outputs.

I messed this up and TestNN.test_MaxPool2d_indices caught me out
on it.  This patch assumes that IndexTensor outputs are not
differentiable.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
As you can see from CI, this patch isn't good enough; I blithely assumed that every Tensor return of any multi-return method was differentiable, but `gesv` breaks the pattern.  I don't have a good sense for where the best place to put differentiability information is. My current plan is to use the occurrence of `grad` in the formula to determine if single return is assumed or not.

ezyang committed:
Hack to detect when only one output is differentiable.

apaszke commented on aten/src/ATen/NativeFunctions.h:
```diff
@@ -112,8 +112,12 @@ type_method_definition_dispatch: at::native::expand
-    throw std::runtime_error("the number of sizes provided must be greater or equal to the "
-                             "number of dimensions in the tensor");
+    std::ostringstream ss;
+    ss << "at::expand(" << self.type() << "{" << self.sizes() << "}, size=" << size
```
Can we not mention any C++ related things in here? These are messages visible to Python users and they don't know what `at::expand` is. Just say what's the problem

ezyang replied:
So, we have a more general problem, where the new ATen code may raise an exception somewhere deep inside an implementation, even though the user visible trace doesn't mention the problem at all. For example, I triggered this error message above by passing an incorrectly shaped gradient out of backwards. Then I got an "expand" error when invoking "backward()". Very perplexing. Inspecting the C++ trace in gdb helped elucidate the situation, but this is not very user friendly.

apaszke replied:
I understand that you can raise a C++ specific error when there's e.g. some ATen invariant that was violated, but in this case there is absolutely no need to mention ATen. It only makes it more confusing for people who don't know that it exists

apaszke commented on tools/autograd/derivatives.yaml:
```diff
@@ -657,9 +660,7 @@
-  grad_output: zeros_like(grad_output)
-  input: zeros_like(input)
-  weight: zeros_like(weight)
+  grad_output, input, weight: prelu_double_backward(grads[0], grads[1], grad_output, input, weight, grad_mask)
```
What?! Please tell me that it's the only case where we have this dummy zero grad... This will silently give people incorrect results! @colesbury? Why are they even marked as implemented if they're broken?

ezyang replied:
There are a lot of other instances of `zeros_like`. They are not wired up to be user visible though. (gradcheck should catch it if we accidentally enable a zeroed grad)

gchanan commented on aten/src/ATen/templates/Tensor.h:
```diff
@@ -129,6 +133,10 @@ struct Tensor : public detail::TensorBase {
   //Tensor * add(Tensor & b);
   ${tensor_method_declarations}
+
+  inline Tensor expand_as(const Tensor & t) const {
```
Any reason these aren't native functions?  They won't end up in Declarations.yaml if you define them here.

ezyang replied:
No reason, I think; I just took the path of least resistance for defining these.

A counter-question: how do we decide whether or not a method goes in `Tensor.h` or is a native function? There are a number of methods in `Tensor.h` and I don't really understand the calculation here.

gchanan replied:
I don't think we should be adding more to Tensor.h; I'd view what is there now as the bootstrap (some of it deals with TH stuff that we don't want to deal with in native functions).

zdevito replied:
The stuff in Tensor.h predates native functions, which is why they are not listed. New stuff should be native functions.  Before we move _everything_ over, we also need to make the native function declaration more concise and make sure the non-dynamic paths (that just drop the method into Tensor.h, for instance) works. @gchanan let me know if you want to chat about how to build out the native function interface.

ezyang committed:
Delete unused defined_if fields.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

colesbury reviewed (commented):
This generally looks good, but the `prelu_double_backward` looks too complicated. For example, the derivative for grad_output should just be something like:

```
grad_output: std::get<0>(prelu_backward(grads[0], input, weight, {true, false}))
```

It's even simpler if we don't fuse the grad_input and grad_weight computations.

I think it's worth taking a step back and figuring out if we should actually be fusing all these backward computations into a single call before implementing convolution

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -85,7 +85,7 @@
 DERIVATIVE_MULTI = CodeTemplate("""\
 if (should_compute_output({ ${idxs} })) {
-  auto output_mask = std::array<bool, ${n}>{
+  auto ${output_mask} = std::array<bool, ${n}>{
```
Yeah, we can leave it as `output_mask`, if the "other" mask is never useful. However, what we should do is error if someone tries to name an input parameter `output_mask`, since it will get shadowed by the `output_mask` we autogenerate.

ezyang replied:
I ended up renaming this to 'grad_input_mask', but doing so uniformly for everything.

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -761,14 +778,34 @@ def emit_body(declaration):
+            # return of the corresponding forward was meant to be
+            # differentiable.  If you actually wanted to differentiate
+            # on the other returns, you would have used 'grads' instead.
+            # This happens in practice with 'gesv'.
```
Exactly; gesv has two outputs. If we treat both outputs of gesv as differentiable, this doesn't work. On the other hand, we need to treat both outputs of prelu_backward as differentiable. So to distinguish between these two cases, we look and see if you used grad or grads in the formula.

apaszke replied:
Can you mark the second output of `gesv` as non-differentiable?

ezyang replied:
We don't have syntax for it, and I couldn't think of any obvious way to extend `derivatives.yaml` to indicate this. Or I guess we could put it in `Declarations.cwrap` but that feels like an abstraction violation.

ezyang committed:
s/thpp/at/

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Delete default parameters from derivatives.yaml.

They don't actually do anything and they're not accurate (many functions
have defaults which we didn't specify here).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Fix parsing problem with std::array<int, 2> (note space)

We are splitting on ', ', but that causes problems when you
have a nested comma.  Quick and dirty fix is to NOT have the
space.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Extra sanity checking for derivatives.yaml versus Declaraitons.yaml

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Fix misnamed generator argument.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
I started working on @colesbury's review comments and got sucked into some more refactoring. The new commits are:

* Delete unused defined_if fields.  …
* s/thpp/at/  …
* Delete default parameters from derivatives.yaml.  …
* Fix parsing problem with std::array<int, 2> (note space)  …
* Extra sanity checking for derivatives.yaml versus Declaraitons.yaml …
* Fix misnamed generator argument. 

The prelu commit also got modified in a pretty substantial way. First, I found a bug in how I was doing `derivatives.yaml`: I added an explicit definition for `prelu` in `derivatives.yaml`, but ACTUALLY preprocess_nn_declarations already gave a definition. I am not sure about the wisdom of these implicit definitions, but in this regime, I shouldn't define a derivatives for 'prelu'. So I deleted it. I then wanted to add a test for whether or not this happened, but I ran afoul this bug: https://github.com/zdevito/ATen/issues/166

The other major change is how I decided to close the loop on `output_mask`. Basically, I really DON'T want the `output_mask` we define in the backwards autogen code to shadow an input named `output_mask` because that is hella confusing. So, I decided to change our local name to `grad_input_mask`, unconditionally. But this triggered a bug in how we do the implicit NN derivatives, so I also fixed that code up too. Now I think everything is nicely consistent.

colesbury commented on aten/src/ATen/NativeFunctions.h:
```diff
@@ -5,11 +5,52 @@
+return: Tensor
+variants: method, function
+type_method_definition_level: base
+type_method_definition_dispatch: at::native::type_as
```
typo: s/type_as/expand_as

colesbury commented on tools/autograd/gen_variable_type.py:
```diff
@@ -449,11 +481,22 @@ def preprocess_nn_functions(declarations):
         actuals = [actual(arg) for arg in bwd['arguments']]
         formula = '{}({})'.format(bwd_name, ', '.join(actuals))
         formula = formula.replace('grad_output', 'grad')
+        # TODO: Ewwwwwww.  This isn't going to work when there are
```
I don't think this comment makes sense. This is only for functions defined by THNN. The formula is created programmatically from the names of the arguments -- it's not written in derivatives.yaml

ezyang replied:
Yeah, you are right!

colesbury reviewed (commented):
Can you post a diff of the generated VariableType.cpp?

colesbury commented:
Other than the typo lgtm

ezyang commented:
VariableType didn't change, but Functions.cpp did:

```
variable_list PreluBackwardBackward::apply(const variable_list& grads) {
  variable_list grad_inputs{3};
  auto grad_output = grad_output_.unpack();
  auto input = input_.unpack();
  auto weight = weight_.unpack();
  if (should_compute_output({ 0, 1, 2 })) {
    auto grad_input_mask = std::array<bool, 3>{
      should_compute_output(0),
      should_compute_output(1),
      should_compute_output(2),
    };
    std::tie(grad_inputs[0], grad_inputs[1], grad_inputs[2]) = prelu_double_backward(grads[0], grads[1], grad_output, input, weight, grad_input_mask);
  }
  ensure_no_aten_scalars(grad_inputs);
  return grad_inputs;
}
```

With the renaming of mask to `grad_input_mask`, the local no longer shadows the member variable named `output_mask`.
